### ISSUE TITLE
[Intermittent] [iOS] Mozilla VPN is crashing

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/mozilla-vpn-client#8390]
**Important details:**

* there are no clear steps for reproducing this -> we have encountered the crash while being logged out from VPN and opening the client after a device restart, or after a fresh install as well, on previous builds;

**Affected versions:**

* 2.18.0 (2.202310192112);

**Tested Platforms:**

* iOS 16.5;

**Prerequisites:**

* Have Mozilla VPN app installed and open;

* Have a Mozilla account with an active subscription or create a new account and subscribe;

* Be signed in; 

**Steps to reproduce:**

1. Attempt to turn ON the VPN;

**Expected result:**

* The VPN is turned ON, there are no issues;

**Actual result:**

* The VPN enters in “Connecting” state and it crashes; 

**Notes:**

* See the attached logs - time of crash - 10:44 / 12:29;

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/VPN-5766)
┆Reporter: Bianca Hidecuti

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Sentry reports, captured from the attached logs:

  * https://mozilla.sentry.io/discover/vpn-client:3537a669-354e-4d2c-64c0-f2abe68a5d2a ( https://mozilla.sentry.io/discover/vpn-client:3537a669-354e-4d2c-64c0-f2abe68a5d2a )
  * https://mozilla.sentry.io/discover/vpn-client:99560288-a3a3-4efb-f29a-be9f2e043281 ( https://mozilla.sentry.io/discover/vpn-client:99560288-a3a3-4efb-f29a-be9f2e043281 )

Sentry report that might be related, because it was reported 2hrs ago on iOS and Valentina Virlics told me she reproduced and sent a report around that time today:

  * https://mozilla.sentry.io/issues/4312881847/?environment=stage&project=4504197915607040&query=is%3Aunresolved&referrer=issue-stream&stream_index=0 ( https://mozilla.sentry.io/issues/4312881847/?environment=stage&project=4504197915607040&query=is%3Aunresolved&referrer=issue-stream&stream_index=0 )

[EDIT] This second one is a crash test. We can ignore it.

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Two things:

  1. This is looking tangential to our old friend QTBUG-115277 ( https://bugreports.qt.io/browse/QTBUG-115277 ), maybe this time we can come up with repro steps for them to re-open this.
  2. It looks like we are not handling the case when the socket doesn’t activate correctly and still trying to send pings to a non-existent socket. I can try to take this on and fix it, depending on how we decide to prioritize it.

With regards to prioritization, question for Valentina Virlics and Bianca Hidecuti. What is the prevalence of this issue? Was it not reproducing on older versions?

I ask, because this is not new code at all. In fact it’s two year old code. Maybe we did something somewhere else that caused this to issue to happen more often… so that would play a part in defining the priority. But if it’s been going on for a while and has the same prevalence in other versions I propose we mark this a medium.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Beatriz Rizental Machado, we encountered crashes more often on iOS, starting with the 2.16 version of Mozilla VPN, each time with other STRs. See https://mozilla-hub.atlassian.net/browse/VPN-4933 ( https://mozilla-hub.atlassian.net/browse/VPN-4933|smart-link ) / https://mozilla-hub.atlassian.net/browse/VPN-5246 ( https://mozilla-hub.atlassian.net/browse/VPN-5246|smart-link ) / https://mozilla-hub.atlassian.net/browse/VPN-5400 ( https://mozilla-hub.atlassian.net/browse/VPN-5400|smart-link ) . 

As these crashes are intermittent and not always reproducing with the same steps, it is hard to say when a crash it is actually no longer happening, or if we are just not able to trigger it anymore. 

During the regression testing we are using the VPN more intensively across all platforms, hence on iOS as well, so if the crash happens rarely we are likely to miss it before the regression period. 

In 2.18 version we have encountered 2 more crashes, before logging this ticket - 1 during the testing of the new onboarding flow and the other one after installing the client (without previous uninstall), so the occurrence of the crash is relatively low.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs → crash time: 14:09 (I also submitted the crash report).

I encountered this after signing out, leaving the app for a few minutes in the background, while on onboarding screen, reopened the app, and the client crashed.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs from iOS 16.5, while using the 2.18 RC. The app crashed after signing out, swipe-kill and reopening it.

Time of crash: 12:14. 

[^12.14 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Latest crash seems to be something different from the preivous ones: https://mozilla.sentry.io/discover/vpn-client:d88f1a6b-9a50-4c34-666f-b036bb2d978b ( https://mozilla.sentry.io/discover/vpn-client:d88f1a6b-9a50-4c34-666f-b036bb2d978b/ ).

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching logs from iOS 16.5 - Mozilla VPN 2.19, after the VPN crashed. Submitted the crash report at 15:04. 

The crash occured when I switched from prod to stage server.

[^15.04mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Raluca Monica Ilban commented:

Attaching logs from iOS 15.6, using build 2.19.0 (2.202311062211). From another device, the iOS device was removed from 'Device not added' screen. After being signed out, the app crashed after clicking Sign in button. Crash report was submitted at 11:19.

[^1119_7Nov_mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching logs from iOS 16.5 - Mozilla VPN 2.19, after the VPN crashed. Submitted the crash report at 09:45.

The crashed occured after the app was opened while the VPN in OFF state (the app was in background over the weekend). 

[^ios 09.45.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs from iOS 17. Submitted the crash report at 13:50. 

[^1.50.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs from iOS 16.5 while using the 2.19 RC. Submitted the crash report at 13:54.

[^13.54mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching logs from iOS 17 as well. Submitted the crash report at 12:00.

[^ios 17 12.00 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

I was able to reproduce this on Mozilla VPN 2.19 build from App store, on iOS 16.5, after the device was removed. 

However, the Crash reporter screen was not displayed, nor the “VPN has crashed” os modal → the app just quit (at 11:18). Attaching logs as well.

[^11-18 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching logs from 2.19.1 using iOS 16.5, while the app has crashed after signing out and navigating through the onboarding screen.

Submitted the crash report at 10:50.

[^10-50 mozillavpn.log]

cc: Beatriz Rizental Machado, if you want to take a look at the logs.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs from iOS, using the 2.19.1 build, after the app crashed while being in Connecting state.

Crash reporter screen was not displayed. Time of crash: 17:49.

[^17-49 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Bianca Hidecuti First, thank you so much for taking the time to add notes to this ticket each time you get iOS crashes. This is extremely useful.

First things first, each one of the crashes in the last comment are different. 

The first one is similar to others we have seen below and I have talked about it and possible mitigations already on https://mozilla-hub.atlassian.net/browse/VPN-5766?focusedCommentId=785970 ( https://mozilla-hub.atlassian.net/browse/VPN-5766?focusedCommentId=785970|smart-link ) . 

The second one I could not identify anything in the logs, the crash is reported to have happened at 18:03, but the logs only go until 18:02 and nothing screams crash at me, so 🤷‍♀️ . There is some stuff in the logs, such as “Adjust is not activatable” which could be issues specifically about the Adjust stuff we are adding in the new dot release. Those are handled issues though, so I do not think they are causing the crashes. Either way looping in Matt Cleinman just in case. Also, I seem to remember we had issues with crashes on swipe kill for iOS before, do you remember something Bianca Hidecuti ?

The third one went to Sentry (https://mozilla.sentry.io/discover/vpn-client:8b449b75-24d8-4078-025f-da749c640b96/ ( https://mozilla.sentry.io/discover/vpn-client:8b449b75-24d8-4078-025f-da749c640b96/ )) which is helpful. Now this one is happening in the Controller inside the activateInternal function, that is a long function and the issue is happening in a QString assignment operator – something that happens a lot of times in the function – so it’s hard to pinpoint exactly where the issue is. I belive it could be in this part ( https://github.com/mozilla-mobile/mozilla-vpn-client/blob/3b34f68b0ed4395dba505876120c684cba470257/src/controller.cpp#L367 ) of the code, because we have seen crashes in it in other platforms (I don’t remember where, but I saw this yesterday with Basti -- tagging him because maybe he does remember). 

On another note, I want to reccomend that we make the priority of this ticket higher or maybe even a blocker for 2.20. The number of crashes found in iOS are really concerning. Regardless of the fact that QA does tend to use the app way more intensely than an average user, the number of crashes in iOS is way higher than any other platforms given those don’t have such an extensive crashes ticket like this one.

My other reccomendation is that we enable Sentry on production iOS. Given it’s only enabled in staging right now, we have no way to know if users are encountering a high number of crashes in iOS. This is a blind spot right now that could be easily addressed – Sentry does not seem to be causing issues in staging. Why have we not enabled it in production?

cc Lesley Norton Santiago Andrigo

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

{quote}Also, I seem to remember we had issues with crashes on swipe kill for iOS before, do you remember something Bianca Hidecuti  {quote}

Beatriz Rizental Machado, yes, in the past we had random crashes when swipe-killing the app on iOS.

**Comment by data-sync-user** [comment]
➤ Lesley Norton commented:

{quote}*My other reccomendation is that we make it a priority to enable Sentry on production iOS.*{quote}

Fixing Sentry and enabling on production iOS are priorities and I’d happily make these blockers. 

{quote}*On another note, I want to reccomend that we make the priority of this ticket higher or maybe even a blocker for 2.20.*{quote}

We need to figure this out asap but - as much as it pains me to say - it doesn’t make sense to block the release unless there is something new or worse in 2.20. I would support however moving the crashes for which we have Sentry logs into separate tickets and making those blockers depending on how often they are reproducing on stage.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching logs from iOS 17.0, using the latest 2.20 build, after the client crashed (after the VPN entered in No signal stated and disconnected). Time of crash: 12:24.

[^14.24 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Valentina Virlics commented:

Attaching logs from iOS 16, using 2.20 (3665 Testfligh ID) build. VPN client was opened and not functional. After tapping a few times the Settings button from the nav bar the app crashed. Time aprox of the crash 15:11.

[^mozillavpn (5).log]
[^MozillaVPN-2024-01-04-151158.ips]

**Comment by data-sync-user** [comment]
➤ Raluca Monica Ilban commented:

Attaching logs from iOS 15.7.6, using the 2.20.0 (2.202401242206) build, after the client crashed (VPN off, device entered in standby; crash after reopening the app ). Aprox time of crash: 10:23.

[^mozillavpn_25Jan.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs after the client crashed using the 2.20 RC on iOS 16.5. “Crash reporter” screen was displayed this time. Time of crash: 17:40.

[^crash ios mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

I believe Bianca’s latest crash is this one: https://mozilla.sentry.io/issues/4680589371/?project=4504197915607040&query=is%3Aunresolved+os%3A"Darwin+22.5.0"&referrer=issue-stream&statsPeriod=7d&stream_index=0 ( https://mozilla.sentry.io/issues/4680589371/?project=4504197915607040&query=is%3Aunresolved+os%3A%22Darwin+22.5.0%22&referrer=issue-stream&statsPeriod=7d&stream_index=0 )

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

We’ve isolated the problem to a specific function (MacOSPingSender::sendPing, which has had occasional crashes for quite a while) but don’t know exactly where the problem lies (from the logs and Sentry crashes we have). We also don’t have repro steps.

I’ve put up a PR ( https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9069 ) to add some logging to this function, to start figuring out where the problem is. Once this PR is merged, we’ll wait to get additional logs from QA, then will hopefully be able to better diagnose and fix this.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Matt Cleinman, please let us know when the PR is merged so we can attach the additional logs. Thanks!

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

Reminder from an old comment here ( https://mozilla-hub.atlassian.net/browse/VPN-5766?focusedCommentId=785970 ), that I’ve investigated the sendPing failure in the past and narrowed it down to a Qt bug (https://bugreports.qt.io/browse/QTBUG-115277 ( https://bugreports.qt.io/browse/QTBUG-115277 )). Could be something else now, but just FYI.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

I saw that, which is part of why I focused in on this. Where is that diagnosis as to why we’re confident it’s that bit? Could we be sending some bad data some function, or trying to pull some bad data to send to it? (That’s why I wanted to iterate on some logging, to see if there is anything common about the data we send in a crash, to try to see if there is some sort of fix.)

**Comment by data-sync-user** [comment]
➤ Beatriz Rizental Machado commented:

No confidence at all 😅 I just mentioned the previous investigation in case it got lost in the long thread and had some use. The issue that led to that Qt bug is old and I was not able to find it, sorry. I remember it was a crash with a bunch of calls to __sendto in a row… Anyways.

Looking at Bianca’s crash and the code for this issue though, makes me realize that Qt bug has nothing to do with it. We are not calling Qt APIs and instead calling sendto directly (https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/platforms/macos/macospingsender.cpp#L86 ( https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/platforms/macos/macospingsender.cpp#L86 )).

What does jump out is the string of Error: Recvmsg failed logs in the latest report posted by Bianca. That means the socket was not ready ( https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/platforms/macos/macospingsender.cpp#L112 ). We should consider adding a log with the errno returned my recvmsg, which might have some more clues. 

Either way, this brings up the possiblity that sendto is crashing because the socket is not ready. (According these docs ( https://linux.die.net/man/2/sendto ) it should still just return an error instead of crash, but those are Linux kernel docs and this is iOS kernel.) Notice that the WindowsPingSender has a check for a valid socket before it sends a ping ( https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/platforms/windows/windowspingsender.cpp#L83-L85 ), we might consider adding a similar check on iOS to prevent this crash.

Maybe we can test the hypothesis that the crash is caused by a faulty socket, by purposefully making it not be ready at the time of sending a message. This could even be an iOS kernel bug??? Given sendto shouldn’t crash 🤔

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

I can’t find documentation on sendto for iOS, though you can see the header user/include/sys/socket.h in Xcode.

Resetting the socket to an invalid number doesn’t seem to crash it, but I haven’t tested more significantly yet.

If worse comes to worse, we could re-write how we do pings in iOS, but hopefully it doesn’t come to this.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

Bianca Hidecuti this logging is now merged in, and will be present in the new daily builds (not the RCs) starting tonight.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Matt Cleinman, I am attaching logs from the latest 2.21 build after the app crashed. Hope it helps. The crashed happened after unlocking the device - the vpn was in off state. 

Crash time: 10:54. 

[^ios 10.54 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Attaching another set of logs after the app has crashed → 11:01. The app crashed after swipe-killing the app while the VPN was OFF.

[^11.01mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

I was able to reproduce this with Mozilla VPN 2.21.0 (2.202402272206), on iOS 17.3.1 while attempting to turn on the VPN.

Crash report was submitted at 11:24. Attaching logs.

cc: Matt Cleinman, in case this is a different crash than the one initially reported.

[^11.24 mozillavpn.log]

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

Looking at latest set of logs. This is logged 1151 times in 70 milliseconds, which probably is what is leading to the crash:

(MacOSPingSender) Error: Recvmsg failed: 18446744073709551615

361 pings were sent in 5.5 seconds of those logs, which also seems odd: (MacOSPingSender) Debug: Sending ping

Focusing back on the recvmsg error. I think there is some overflow issue. Because Rcvmsg failed is only logged when rc <= 0, and in this case rc is then reported as 18446744073709551615. From the limited documentation ( https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/recvmsg.2.html ), it looks like errors will only return -1 and 0. 

And so this is probably some sort of max integer issue. It seems like the max ssize_t is platform specific. This bug could be one of two things:

  1. Recvmsg is returning 1. We’re erroring out, as we should. But then when printing it in logs, we change that 1 to 18446744073709551615
  2. Recvmsg is actually returning 18446744073709551615. Our code is mistakenly translating that as -1.

I'd say option 1 is waaay more likely, especially as we’re getting the exact translation of -1.

Going to talk this through next week w/ the team, as others know more about this area than I do.

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

This is almost definitely iOS killing the socket on us when app is backgrounded.

Fairly consistent repro steps: 

  1. VPN is off
  2. Cold launch of app
  3. In about a second or two, press power button to turn screen off
  4. In 5-10 seconds, unlock phone (this part seemed inconsistent - the crash doesn’t happen if it’s only a couple seconds)
  5. Crash happens

My guess is that this also happens if you switch to another app, but can’t be confident.

With these repro steps (informed by some research on iOS and sockets), Naomi Kirby put together a fix. PR will be up shortly.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

We will verifiy this when a new build which contains the fix will be available in Testflight - currently the latest builds from Testflight is 3 days old - cc: Matt Cleinman

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Verified this on Mozilla VPN 2.21 (#3748 from Testflight) using iOS 17.3.1 / iOS 15.7 and we were unable to trigger the crash in the following scenarios: 

  * while the VPN is in “Connecting..” state after making an IAP subscription;
  * after unlocking the device while the VPN was in OFF state and the device in stand-by for several minutes;
  * after removing the device;
  * after swipe-killing the app while the VPN is ON/OFF;

We would like to keep this open for the next few days to make sure that the crash is indeed no longer happening. 

cc: Raluca Monica Ilban / Valentina Virlics

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

Looking at these logs, it seems to be a different crash. Can you create a new ticket for it?

And so I better understand, what happened? It was running fine, you swipe killed the app, you re-opened the app, and then it immediately crashed? Did it show the main screen first, or anything like that?

**Comment by data-sync-user** [comment]
➤ Raluca Monica Ilban commented:

Matt Cleinman , here is the new ticket created for the crash mentioned by Bianca Hidecuti in the previous comment: https://mozilla-hub.atlassian.net/browse/VPN-6243 ( https://mozilla-hub.atlassian.net/browse/VPN-6243|smart-link )

**Comment by data-sync-user** [comment]
➤ Matt Cleinman commented:

With a new ticket for https://mozilla-hub.atlassian.net/browse/VPN-6243 ( https://mozilla-hub.atlassian.net/browse/VPN-6243|smart-link ), I’m moving this back to merged.
